### ISSUE TITLE
DriverIOError: Failed to create a new (non existing yet) Shapefile on the disk

### ISSUE THREAD

**Issue by swiss-knight** [Toblerity/Fiona#871]
Hi devs,

I encountered a new problem this morning after having updated fiona.

Here's the log:
```
Traceback (most recent call last):

  File "<ipython-input-71-8a0d31d12aff>", line 16, in <module>
    'w', 'ESRI Shapefile', schemaf) as f:

  File "/usr/local/lib/python3.6/dist-packages/fiona/env.py", line 407, in wrapper
    return f(*args, **kwargs)

  File "/usr/local/lib/python3.6/dist-packages/fiona/__init__.py", line 264, in open
    **kwargs)

  File "/usr/local/lib/python3.6/dist-packages/fiona/collection.py", line 157, in __init__
    self.session.start(self, **kwargs)

  File "fiona/ogrext.pyx", line 1066, in fiona.ogrext.WritingSession.start

DriverIOError: Failed to create file  /media/username/data/TEMP/19.shp: No such file or directory

```

## Expected behavior and actual behavior.
I am writing a new Shapefile from data, which is an instance of a `shapely.geometry.polygon.Polygon`.    
The Shapefile obviously doesn't exist on the disk yet.    
It was working before the update.

## Steps to reproduce the problem.
```
import os
import fiona
from shapely.geometry import Polygon, mapping

base_directory = '/media/username/data/TEMP/' # feel free to adapt this path
k=19
schemaf = {
    'geometry': 'Polygon', 
    'properties': {'id': 'int'},
}
my_shapely_simple_polygon = Polygon([(0, 0), (1, 1), (1, 0)])
my_shapefile = os.path.join(base_directory, str(k)+'.shp')

# Write the data to a Shapefile:
with fiona.open(my_shapefile, 
                'w', 'ESRI Shapefile', schemaf) as f:
    f.write({'geometry': mapping(my_shapely_simple_polygon), # instance of shapely.geometry.polygon.Polygon
             'properties': {'id': k},})
```

## Operating system

Ubuntu 18.04 64 bit `5.3.0-42-generic x86_64 GNU/Linux`

## Fiona and GDAL version and provenance    
`GDAL 2.4.0, released 2018/12/14`    
`Fiona                         1.8.13.post1`     
`Shapely                       1.7.0`    
 
In python3:
```
>>> from osgeo import gdal
>>> gdal.__version__
'2.4.0'     

>>> import fiona
>>> fiona.__version__
'1.8.13.post1'    

>>> import shapely
>>> shapely.__version__
'1.7.0'    

```
Python3 itself:
```
Python 3.6.9 (default, Nov  7 2019, 10:44:02) 
[GCC 8.3.0] on linux.
```

Thanks!

**Comment by rbuffat** [comment]
@swiss-knight 

Does it work if you remove the white space in front of the path?
``base_directory = ' /media/username/data/TEMP/' # feel free to adapt this path``

**Comment by swiss-knight** [comment]
> @swiss-knight
> 
> Does it work if you remove the white space in front of the path?
> `base_directory = ' /media/username/data/TEMP/' # feel free to adapt this path`

Fixed.     
It doesn't change the behavior of the issue.

**Comment by rbuffat** [comment]
```
import os
import fiona
from shapely.geometry import Polygon, mapping

base_directory = '/tmp' # feel free to adapt this path
k=19
schemaf = {
    'geometry': 'Polygon', 
    'properties': {'id': 'int'},
}
my_shapely_simple_polygon = Polygon([(0, 0), (1, 1), (1, 0)])
my_shapefile = os.path.join(base_directory, str(k)+'.shp')

# Write the data to a Shapefile:
with fiona.open(my_shapefile, 
                'w', 'ESRI Shapefile', schemaf) as f:
    f.write({'geometry': mapping(my_shapely_simple_polygon), # instance of shapely.geometry.polygon.Polygon
             'properties': {'id': k},})
```

works for me with Fiona 1.8.13.post1. 

Could you test if this works also for you? If yes, can you check if the directory /media/username/data/TEMP actually exists and you have write permissions on it?

**Comment by swiss-knight** [comment]
Indeed, this also works for me.

Here is my rights on the already existing folder on my disk:

```
$ ls -larth /media/username/Data/ | grep TEMP
drwxr-xr-x   2 username username 4.0K avr  1 13:49 TEMP
```

Strange...

**Comment by rbuffat** [comment]
I'm currently not sure what causes this. The obvious cause would have been the directory to write in does not exist or writing is not possible.  /media is normally used for removable media, thus it could be that there is an issue with that.

Does it work e.g. with the GPKG driver?

The error happens probably within the GDAL Shapefile driver: https://github.com/OSGeo/gdal/blob/beb2622c94ea36c3f392de3ab00c960108815353/gdal/ogr/ogrsf_frmts/shape/shpopen.c#L1301

Do you use the ubuntu package or have you Fiona installed with pip?
You could try to downgrade Fiona to the previous version and check if the gdal version Fiona uses has changed with ``fiona.get_gdal_release_name()``. If this behavior depends on the version of gdal, this could be a hint.

**Comment by swiss-knight** [comment]
*earasing all what I've written because I've been surprised it was working with the previous version of fiona, and then, that it was **always** working once back to the latest version (where it didn't worked before)*

So, I investigated a little more from that point, to figure out I honestly made a mistake.    
I'll try to explain what happened so that it may, hopefully, help others...    

Somewhere before the short snippet I pasted in my script, I had this piece of code;

```
base_directory = os.path.join(project_directory, 'TEMP')
(...)
try:
    if not os.path.isdir(base_directory):
        os.mkdir(base_directory)
except:
    pass
```

Then I used the `base_directory` as the root directory where to store the Shapefile.

It went all fine.

But I made a recent upgrade to split different types of data into two new sub-folders;

```
base_directory = os.path.join(project_directory, 'TEMP', 'sub_folder_1')
(...)
try:
    if not os.path.isdir(base_directory):
        os.makedirs(base_directory)
except:
    pass
```

Then the fiona error was raised!

And indeed, what should have been raised here is this, but it was muted because of the `try` `catch`:
```
os.mkdir(base_directory)

Traceback (most recent call last):

  File "<ipython-input-37-452d89490330>", line 1, in <module>
    os.mkdir(base_directory)

FileNotFoundError: [Errno 2] No such file or directory: '/media/username/Data/TEMP/sub_folder_1/'
```

This is obviously because `os.mkdir()` does **not** handle building recursively sub-directories, as the `-p` flag of the real Ubuntu `mkdir` command would have done.     
So prior to the change, it was able to make that `TEMP` directory because it was at the first hierarchical level from the existing project directory.    
And after the change, I asked this method to directly build a level 2 sub-directory, namely, `TEMP/sub_folder_1`, which is not possible.    
Instead, I have to use `os.makedirs()` which handles such recursive cases!

I'm really sorry for having making you guys lose of your precious time. :( I feel ashamed.
